ir/src/boundary_constraints.rs

```diff
@@ -0,0 +1,76 @@
+use std::collections::BTreeMap;
+
+use super::{Expr, Identifier, SemanticError, TraceColumns};
+use parser::ast;
+
+// BOUNDARY CONSTRAINTS
+// ================================================================================================
+
+/// The map of constraints to be applied at the specified boundary
+#[derive(Default, Debug)]
+pub(crate) struct BoundaryConstraints {
+    /// The boundary constraints to be applied at the first row of the trace, with the trace column
+    /// index as the key, and the expression as the value.
+    first: BTreeMap<usize, Expr>,
+    /// The boundary constraints to be applied at the last row of the trace, with the trace column
+    /// index as the key, and the expression as the value.
+    last: BTreeMap<usize, Expr>,
```

Comment:
Yes, I agree in theory, and I actually started by doing that in a few different ways. I changed it back as a simplification, because the AST `Expr` and `Identifier` are actually all we need (caveat below), so it seems redundant to map it to an identical struct. One thing I thought about is to define the `Expr` and `Identifier` in the IR and then import them to the parser and use them there. I didn't do it in the first pass because the IR was already referencing the parser but not vice versa.

Technically the `Expr` for boundary and transition constraints should be different, since boundary constraints can't reference trace columns. We have 2 options for this
1. define them differently at the AST stage so these errors are raised during parsing
2. define a BoundaryExpr struct in the IR, convert from the `ast::Expr` and raise an error at this stage.

I think option 1 is better - but this would still leave us with `ast::Expr` exactly equivalent to what we want for boundary expressions in the IR.